### PR TITLE
fix: add prompt field in RAGTool

### LINKED ISSUE #132: [BUG] RAGTool is not able to give response.
**What is the bug?**
When asking agent on the questions: `Can you tell me what's the population increase of Seattle from 2021 to 2023 by using RAGTool?`, the agent will gives an error of
```
2024-01-16 15:28:58 | java.lang.IllegalArgumentException: Some parameter placeholder not filled in payload: prompt
-- | --
2024-01-16 15:28:58 | at org.opensearch.ml.common.connector.Connector.validatePayload(Connector.java:88) ~[opensearch-ml-common-3.0.0.0-SNAPSHOT.jar:?]
2024-01-16 15:28:58 | at org.opensearch.ml.engine.algorithms.remote.RemoteConnectorExecutor.preparePayloadAndInvokeRemoteModel(RemoteConnectorExecutor.java:118) ~[opensearch-ml-algorithms-3.0.0.0-SNAPSHOT.jar:?]
2024-01-16 15:28:58 | at org.opensearch.ml.engine.algorithms.remote.RemoteConnectorExecutor.executePredict(RemoteConnectorExecutor.java:74) ~[opensearch-ml-algorithms-3.0.0.0-SNAPSHOT.jar:?]
2024-01-16 15:28:58 | at org.opensearch.ml.engine.algorithms.remote.RemoteModel.predict(RemoteModel.java:61) ~[opensearch-ml-algorithms-3.0.0.0-SNAPSHOT.jar:?]
```

**How can one reproduce the bug?**
1. Go to Chatbot
2. Click on Chatbot button
3. Start a new conversation and ask `Can you tell me what's the population increase of Seattle from 2021 to 2023 by using RAGTool?`

**What is the expected behavior?**
It should response without error

**What is your host/environment?**
 - OS: Linux
 - Version 3.0.0
 - Plugins

**Do you have any screenshots?**
image

**Do you have any additional context?**
Add any other context about the problem.

### PR DESCRIPTION
**Pull request by SuZhou-Joe**
### Description
[Describe what this change achieves]
 
### Issues Resolved
#132 
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by xinyual** [comment]
So will we have a default prompt for RAG tool?

**Comment by SuZhou-Joe** [comment]
#132

File: src/main/java/org/opensearch/agent/tools/RAGTool.java [hunk 1]
```diff
@@ -156,6 +160,7 @@ public <T> void run(Map<String, String> parameters, ActionListener<T> listener)
 
             Map<String, String> tmpParameters = new HashMap<>();
             tmpParameters.putAll(parameters);
```
Comment (mingshl) [inline src/main/java/org/opensearch/agent/tools/RAGTool.java:162]: here is where the `INPUT_FIELD` get feeds into the llm tool parameters

File: src/main/java/org/opensearch/agent/tools/RAGTool.java [hunk 2]
```diff
@@ -256,6 +261,7 @@ public RAGTool create(Map<String, Object> params) {
             String[] sourceFields = gson.fromJson((String) params.get(SOURCE_FIELD), String[].class);
             String inferenceModelId = (String) params.get(INFERENCE_MODEL_ID_FIELD);
             Integer docSize = params.containsKey(DOC_SIZE_FIELD) ? Integer.parseInt((String) params.get(DOC_SIZE_FIELD)) : 2;
+            String prompt = (String) params.get(PROMPT_FIELD);
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/main/java/org/opensearch/agent/tools/RAGTool.java:264]: Actually in tool.run, tool can only get parameters.run, the prompt config when registering the tool can only be accessible in create method.

**Comment by mingshl** [comment]
> So will we have a default prompt for RAG tool?

if the prompt is going to add in when register a tool, we should either check if empty input then throw exception, or give the prompt a default.

**Review by mingshl — COMMENTED** [review]
```
if (prompt.trim().isEmpty() || prompt == null) {
     throw new IllegalArgumentException("prompt cannot be empty");
}
```

**Comment by zhichao-aws** [comment]
Can we set a default prompt for claude model? like https://github.com/opensearch-project/skills/pull/125

**Comment by SuZhou-Joe** [comment]
> So will we have a default prompt for RAG tool?

I once heard that we should never put prompt in our code, it should be a config from user. Will add a validation in Tool.

**Comment by SuZhou-Joe** [comment]
Based on discussion with @mingshl , we decide to add a validation on prompt field instead of adding a default prompt as the prompt may contain claude-specific template like: `Human: turn`, which will introduce confusion here.
@mingshl @xinyual @zhichao-aws Could you please help to review and get it merged?

**Comment by mingshl** [comment]
Hi @SuZhou-Joe , 

IMO, it's better the the `prompt` should be carried through the tool id, not the RAGTool interface. so I took a look at [MLModelTool](https://github.com/opensearch-project/ml-commons/blob/main/ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/MLModelTool.java) which is similar to RAGTool that takes a prompt when register the tool and ask llm model a question in a request, MLModelTool doesn't need the prompt in the constructor, the prompt should added through [LLMSpec](https://github.com/opensearch-project/ml-commons/blob/main/common/src/main/java/org/opensearch/ml/common/agent/LLMSpec.java) and then in every MLPredictionTaskRequest, it will carry the prompt to use in model prediction. 

based on the previous testing on ml-commons feature branch, the RAGToold without the prompt in the constructor was working, I would like to double check if the setting of the RAGTool might lead to issue https://github.com/opensearch-project/skills/issues/132

**Comment by xinyual** [comment]
> > So will we have a default prompt for RAG tool?
> 
> if the prompt is going to add in when register a tool, we should either check if empty prompt then throw exception, or give the prompt a default.

Yes. I think we could add a default prompt if the prompt is empty like https://github.com/opensearch-project/skills/pull/125

**Pull request closed without merging**